src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1239,96 +1242,126 @@ private bool IsRefStruct(out bool isRefFollowedByStructOrPartialStruct, out bool
             return isTypeDeclaration && nextToken.Kind == SyntaxKind.StructKeyword;
         }
 
-        private static bool CanFollowTypeName(SyntaxKind kind)
+        private static bool IsContextualModifier(SyntaxToken token)
         {
-            switch (kind)
+            switch (token.ContextualKind)
             {
-                case SyntaxKind.DotToken:                   // partial.     qualified
-                case SyntaxKind.ColonColonToken:            // partial::    alias
-                case SyntaxKind.LessThanToken:              // partial<     generic type
-                case SyntaxKind.OpenBracketToken:           // partial[     array
+                case SyntaxKind.PartialKeyword:
+                case SyntaxKind.AsyncKeyword:
                     return true;
             }
 
             return false;
         }
 
-        private static bool CanFollowMemberName(SyntaxKind kind)
+        // Returns true if the current token is probably a modifier.
+        // To avoid further lookahead, caller is responsible to disambiguate some edge cases.
```

Comment:
i personally don't like this.  why not just have a single place that checks "Is this partial a partial modifier or not"?  :-/

Reply:
note:  i think that's what you're doing.  however, if so, i would make this function a local function helper of the main scanning function you have.  but, that's less important.

Reply:
>Is this partial a partial modifier or not

Something like `IsTrueModifier` requires the same or more amount of analysis as ScanPartial method which is not necessary here. we optimistically look past those suspicious modifiers and later we disambiguate based on just the last modifier. (that's exactly what you suggested earlier).

Reply:
You're right.  I think my only concern is that this is just a random helper hanging out in the middle of nowhere.  If it really is only used by your ScanPartial... method, i would prefer it be isolated only to that. in which case, it would make more sense (to me at least).  I retract my original complaint as this makes more sense now.  But i would like to see if it could be clearer with a little bit of reorganization.

Reply:
> I think my only concern is that this is just a random helper hanging out

I think this will be used for async parsing as well as part of https://github.com/dotnet/roslyn/pull/25704.

Reply:
That makes sense.  LGTM!  Thanks!